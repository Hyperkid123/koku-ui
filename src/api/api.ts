import axios, { AxiosRequestConfig } from 'axios';

export interface PagedResponse<D = any> {
  count: number;
  previous: string;
  next: string;
  results: D[];
}

export function initApi({ version }: { version: string }) {
  axios.defaults.baseURL = `/r/insights/platform/cost-management/api/${version}/`;
  axios.interceptors.request.use(authInterceptor);
}

export function authInterceptor(
  reqConfig: AxiosRequestConfig
): AxiosRequestConfig {
  return {
    ...reqConfig,
    headers: {
      ...reqConfig.headers,
    },
  };
}
