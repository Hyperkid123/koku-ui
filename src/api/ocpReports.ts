import axios from 'axios';
import { Omit } from 'react-redux';

export interface OcpDatum {
  value: number;
  units: string;
}

export interface OcpReportValue {
  capacity?: OcpDatum;
  cost?: OcpDatum;
  cluster?: string;
  cluster_alias?: string;
  count?: OcpDatum;
  date: string;
  delta_percent?: number;
  delta_value?: number;
  derived_cost?: OcpDatum;
  infrastructure_cost?: OcpDatum;
  limit?: OcpDatum;
  node?: string;
  project?: string;
  request?: OcpDatum;
  usage: OcpDatum;
}

export interface GroupByClusterData extends Omit<OcpReportData, 'clusters'> {
  service: string;
}

export interface GroupByNodeData extends Omit<OcpReportData, 'nodes'> {
  region: string;
}

export interface GroupByProjectData extends Omit<OcpReportData, 'projects'> {
  account: string;
}

export interface OcpReportData {
  clusters?: GroupByClusterData[];
  date?: string;
  delta_percent?: number;
  delta_value?: number;
  nodes?: GroupByNodeData[];
  projects?: GroupByProjectData[];
  values?: OcpReportValue[];
}

export interface OcpReportMeta {
  delta?: {
    percent: number;
    value: number;
  };
  group_by?: {
    [group: string]: string[];
  };
  order_by?: {
    [order: string]: string;
  };
  filter?: {
    [filter: string]: any;
  };
  total?: {
    capacity?: OcpDatum;
    cost: OcpDatum;
    derived_cost: OcpDatum;
    infrastructure_cost: OcpDatum;
    limit?: OcpDatum;
    request?: OcpDatum;
    usage?: OcpDatum;
  };
  count: number;
}

export interface OcpReportLinks {
  first: string;
  previous?: string;
  next?: string;
  last: string;
}

export interface OcpReport {
  meta: OcpReportMeta;
  links: OcpReportLinks;
  data: OcpReportData[];
}

export const enum OcpReportType {
  cost = 'cost',
  cpu = 'cpu',
  memory = 'memory',
  tag = 'tag',
  volume = 'volume',
}

export const OcpReportTypePaths: Record<OcpReportType, string> = {
  [OcpReportType.cost]: 'reports/openshift/costs/',
  [OcpReportType.cpu]: 'reports/openshift/compute/',
  [OcpReportType.memory]: 'reports/openshift/memory/',
  [OcpReportType.tag]: 'tags/openshift/',
  [OcpReportType.volume]: 'reports/openshift/volumes/',
};

export function runReport(reportType: OcpReportType, query: string) {
  const path = OcpReportTypePaths[reportType];
  return axios.get<OcpReport>(`${path}?${query}`);
}
