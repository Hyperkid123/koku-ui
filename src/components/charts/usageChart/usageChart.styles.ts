import { chart_color_green_100 } from '@patternfly/react-tokens/dist/esm/chart_color_green_100';
import { chart_color_green_200 } from '@patternfly/react-tokens/dist/esm/chart_color_green_200';
import { chart_color_green_300 } from '@patternfly/react-tokens/dist/esm/chart_color_green_300';
import { chart_color_green_400 } from '@patternfly/react-tokens/dist/esm/chart_color_green_400';
import { chart_color_green_500 } from '@patternfly/react-tokens/dist/esm/chart_color_green_500';
import { global_disabled_color_200 } from '@patternfly/react-tokens/dist/esm/global_disabled_color_200';
import { global_spacer_lg } from '@patternfly/react-tokens/dist/esm/global_spacer_lg';

export const chartStyles = {
  chartContainer: {
    marginTop: global_spacer_lg.value,
  },
  currentRequestData: {
    data: {
      fill: 'none',
      stroke: '#88D080',
      strokeDasharray: '3,3',
    },
  },
  currentUsageData: {
    data: {
      fill: 'none',
      stroke: '#A2DA9C',
    },
  },
  itemsPerRow: 2,
  // See: https://github.com/project-koku/koku-ui/issues/241
  legendColorScale: [
    global_disabled_color_200.value,
    chart_color_green_100.value,
    global_disabled_color_200.value,
    chart_color_green_200.value,
  ],
  previousRequestData: {
    data: {
      fill: 'none',
      stroke: global_disabled_color_200.value,
      strokeDasharray: '3,3',
    },
  },
  previousUsageData: {
    data: {
      fill: 'none',
      stroke: global_disabled_color_200.value,
    },
  },
  // See: https://github.com/project-koku/koku-ui/issues/241
  currentColorScale: [
    chart_color_green_100.value,
    chart_color_green_200.value,
    chart_color_green_300.value,
    chart_color_green_400.value,
    chart_color_green_500.value,
  ],
  // TBD: No grey scale, yet
  previousColorScale: [
    global_disabled_color_200.value,
    global_disabled_color_200.value,
  ],

  yAxis: {
    axisLabel: {
      padding: 15,
    },
    grid: {
      stroke: 'none',
    },
    ticks: {
      stroke: 'none',
    },
    tickLabels: {
      fontSize: 0,
    },
  },
  xAxis: {
    axisLabel: {
      padding: 40,
    },
    grid: {
      stroke: 'none',
    },
    ticks: {
      stroke: 'none',
    },
  },
};
