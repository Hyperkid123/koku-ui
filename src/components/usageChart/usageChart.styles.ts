import { StyleSheet } from '@patternfly/react-styles';
import {
  global_disabled_color_100,
  global_disabled_color_200,
  global_FontFamily_sans_serif,
  global_FontSize_md,
  global_spacer_lg,
} from '@patternfly/react-tokens';
import { VictoryStyleInterface } from 'victory';

export const chartStyles = {
  axis: {
    axisLabel: {
      padding: 15,
    },
    grid: {
      stroke: 'none',
    },
    ticks: {
      stroke: 'none',
    },
    tickLabels: {
      fontSize: 0,
    },
  } as VictoryStyleInterface,
  currentRequestData: {
    data: {
      fill: 'none',
      stroke: '#88D080',
      strokeDasharray: '3,3',
    },
  } as VictoryStyleInterface,
  currentUsageData: {
    data: {
      fill: 'none',
      stroke: '#A2DA9C',
    },
  } as VictoryStyleInterface,
  legend: {
    labels: {
      fontFamily: global_FontFamily_sans_serif.value,
      fontSize: 12,
    },
  },
  previousRequestData: {
    data: {
      fill: 'none',
      stroke: global_disabled_color_200.value,
      strokeDasharray: '3,3',
    },
  } as VictoryStyleInterface,
  previousUsageData: {
    data: {
      fill: 'none',
      stroke: global_disabled_color_200.value,
    },
  } as VictoryStyleInterface,
  // See: https://github.com/project-koku/koku-ui/issues/241
  currentColorScale: ['#A2DA9C', '#88D080', '#6EC664', '#519149', '#3C6C37'],
  // TBD: No grey scale, yet
  previousColorScale: [
    global_disabled_color_200.value,
    global_disabled_color_100.value,
  ],
};

export const styles = StyleSheet.create({
  chartContainer: {
    ':not(foo) svg': {
      overflow: 'visible',
    },
  },
  legend: {
    display: 'inline-block',
    fontSize: global_FontSize_md.value,
    marginBottom: global_spacer_lg.value,
    minWidth: '175px',
    width: '50%',
  },
});
