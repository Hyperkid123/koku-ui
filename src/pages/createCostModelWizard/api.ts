import { fetchCostModels } from 'api/costModels';
import { fetchProviders } from 'api/providers';

export const fetchSources = ({ type, page, perPage, query }) => {
  const offset = (page - 1) * perPage;
  const limit = perPage;
  const queryParam = Object.keys(query).reduce(
    (acc, cur) =>
      acc ? `${acc}&${cur}=${query[cur]}` : `${cur}=${query[cur]}`,
    ''
  );
  return fetchProviders(
    `type=${type}&limit=${limit}&offset=${offset}&${queryParam}`
  )
    .then(sources => sources.data.data)
    .then(sources => {
      return fetchCostModels().then(cms => ({
        costmodels: cms.data.data,
        sources,
      }));
    })
    .then(({ costmodels, sources }) => {
      const cmsHash = costmodels.reduce((acc, curr) => {
        curr.provider_uuids.forEach(provider_uuid => {
          acc[provider_uuid] = curr.name;
        });
        return acc;
      }, {});
      return sources.map(src => ({
        name: src.name,
        costmodel: cmsHash[src.uuid],
        uuid: src.uuid,
        selected: false,
      }));
    });
};
