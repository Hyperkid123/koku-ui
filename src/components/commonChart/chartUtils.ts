import { AwsReport } from 'api/awsReports';
import { OcpReport } from 'api/ocpReports';
import format from 'date-fns/format';
import getDate from 'date-fns/get_date';
import startOfMonth from 'date-fns/start_of_month';
import { FormatOptions, ValueFormatter } from 'utils/formatValue';
import {
  ComputedAwsReportItem,
  getComputedAwsReportItems,
  getIdKeyForGroupBy,
} from 'utils/getComputedAwsReportItems';
import {
  ComputedOcpReportItem,
  getComputedOcpReportItems,
} from 'utils/getComputedOcpReportItems';
import { SortDirection } from 'utils/sort';

export interface ChartDatum {
  x: string | number;
  y: number;
  key: string | number;
  name?: string | number;
  units: string;
}

export const enum ChartType {
  rolling,
  daily,
  monthly,
}

export function isAwsReport(
  report: AwsReport | OcpReport
): report is AwsReport {
  const groupById = report ? getIdKeyForGroupBy(report.group_by) : 'date'; // default key
  return groupById !== 'date';
}

export function transformReport(
  report: AwsReport | OcpReport,
  type: ChartType = ChartType.daily,
  key: any = 'date'
): ChartDatum[] {
  if (!report) {
    return [];
  }
  const items = {
    report,
    idKey: key,
    sortKey: 'id',
    sortDirection: SortDirection.desc,
  } as any;
  const computedItems = isAwsReport(report)
    ? getComputedAwsReportItems(items)
    : getComputedOcpReportItems(items);
  if (type === ChartType.daily) {
    return computedItems.map(i => createDatum(i.total, i, key));
  }

  if (type === ChartType.monthly) {
    return computedItems.map(i => createDatum(i.total, i, key));
  }

  return computedItems.reduce<ChartDatum[]>((acc, d) => {
    const prevValue = acc.length ? acc[acc.length - 1].y : 0;
    return [...acc, createDatum(prevValue + d.total, d, key)];
  }, []);
}

export function createDatum(
  value: number,
  computedItem: ComputedAwsReportItem | ComputedOcpReportItem,
  idKey = 'date'
): ChartDatum {
  const xVal = idKey === 'date' ? getDate(computedItem.id) : computedItem.label;
  return {
    x: xVal,
    y: parseFloat(value.toFixed(2)),
    key: computedItem.id,
    name: computedItem.id,
    units: computedItem.units,
  };
}

export function getDatumDateRange(datums: ChartDatum[]): [Date, Date] {
  if (!datums.length) {
    const today = new Date();
    const firstOfMonth = startOfMonth(today);
    return [firstOfMonth, today];
  }

  const start = new Date(datums[0].key + 'T00:00:00');
  const end = new Date(datums[datums.length - 1].key + 'T00:00:00');
  return [start, end];
}

export function getDateRangeString(datums: ChartDatum[]) {
  const [start, end] = getDatumDateRange(datums);
  const monthName = format(start, 'MMM');
  const startDate = getDate(start);
  const endDate = getDate(end);
  return `${monthName} ${getDate(start)}${
    startDate !== endDate ? ` - ${endDate}` : ''
  }`;
}

export function getTooltipLabel(
  datum: ChartDatum,
  formatValue: ValueFormatter,
  formatOptions?: FormatOptions,
  idKey: any = 'date'
) {
  if (!datum.key) {
    return '';
  }
  if (idKey === 'date') {
    const date = format(datum.key, 'MMM D YYYY');
    return `${date}: ${formatValue(datum.y, datum.units, formatOptions)}`;
  }
  return datum.key.toString();
}
