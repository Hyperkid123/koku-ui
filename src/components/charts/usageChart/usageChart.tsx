import 'components/charts/common/charts-common.scss';

import {
  Chart,
  ChartArea,
  ChartAxis,
  ChartLegend,
  ChartLegendTooltip,
  createContainer,
  getInteractiveLegendEvents,
  getInteractiveLegendItemStyles,
} from '@patternfly/react-charts';
import { Title } from '@patternfly/react-core';
import { default as ChartTheme } from 'components/charts/chartTheme';
import {
  getDateRange,
  getMaxMinValues,
  getTooltipContent,
  getUsageRangeString,
} from 'components/charts/common/chartDatumUtils';
import getDate from 'date-fns/get_date';
import i18next from 'i18next';
import React from 'react';
import { FormatOptions, ValueFormatter } from 'utils/formatValue';
import { DomainTuple, VictoryStyleInterface } from 'victory-core';

import { chartStyles } from './usageChart.styles';

interface UsageChartProps {
  adjustContainerHeight?: boolean;
  containerHeight?: number;
  currentRequestData?: any;
  currentUsageData: any;
  formatDatumValue?: ValueFormatter;
  formatDatumOptions?: FormatOptions;
  height?: number;
  legendItemsPerRow?: number;
  padding?: any;
  previousRequestData?: any;
  previousUsageData?: any;
  title?: string;
}

interface UsageChartData {
  name?: string;
}

interface UsageChartLegendItem {
  name?: string;
  symbol?: any;
  tooltip?: string;
}

interface UsageChartSeries {
  childName?: string;
  data?: [UsageChartData];
  legendItem?: UsageChartLegendItem;
  style?: VictoryStyleInterface;
}

interface State {
  cursorVoronoiContainer?: any;
  hiddenSeries: Set<number>;
  series?: UsageChartSeries[];
  width: number;
}

class UsageChart extends React.Component<UsageChartProps, State> {
  private containerRef = React.createRef<HTMLDivElement>();
  public navToggle: any;
  public state: State = {
    hiddenSeries: new Set(),
    width: 0,
  };

  public componentDidMount() {
    setTimeout(() => {
      if (this.containerRef.current) {
        this.setState({ width: this.containerRef.current.clientWidth });
      }
      window.addEventListener('resize', this.handleResize);
      this.navToggle = insights.chrome.on('NAVIGATION_TOGGLE', this.handleNavToggle);
    });
    this.initDatum();
  }

  public componentDidUpdate(prevProps: UsageChartProps) {
    if (
      prevProps.currentRequestData !== this.props.currentRequestData ||
      prevProps.currentUsageData !== this.props.currentUsageData ||
      prevProps.previousRequestData !== this.props.previousRequestData ||
      prevProps.previousUsageData !== this.props.previousUsageData
    ) {
      this.initDatum();
    }
  }

  public componentWillUnmount() {
    window.removeEventListener('resize', this.handleResize);
    if (this.navToggle) {
      this.navToggle();
    }
  }

  private initDatum = () => {
    const { currentRequestData, currentUsageData, previousRequestData, previousUsageData } = this.props;

    const usageKey = 'chart.usage_legend_label';
    const usageTooltipKey = 'chart.usage_legend_tooltip';
    const requestKey = 'chart.requests_legend_label';
    const requestTooltipKey = 'chart.requests_legend_tooltip';

    // Show all legends, regardless of length -- https://github.com/project-koku/koku-ui/issues/248

    const series: UsageChartSeries[] = [
      {
        childName: 'previousUsage',
        data: previousUsageData,
        legendItem: {
          name: getUsageRangeString(previousUsageData, usageKey, true, true, 1),
          symbol: {
            fill: chartStyles.legendColorScale[0],
            type: 'minus',
          },
          tooltip: getUsageRangeString(previousUsageData, usageTooltipKey, false, false, 1),
        },
        style: chartStyles.previousUsageData,
      },
      {
        childName: 'currentUsage',
        data: currentUsageData,
        legendItem: {
          name: getUsageRangeString(currentUsageData, usageKey, true, false),
          symbol: {
            fill: chartStyles.legendColorScale[1],
            type: 'minus',
          },
          tooltip: getUsageRangeString(currentUsageData, usageTooltipKey, false, false),
        },
        style: chartStyles.currentUsageData,
      },
      {
        childName: 'previousRequest',
        data: previousRequestData,
        legendItem: {
          name: getUsageRangeString(previousRequestData, requestKey, true, true, 1),
          symbol: {
            fill: chartStyles.legendColorScale[2],
            type: 'dash',
          },
          tooltip: getUsageRangeString(previousRequestData, requestTooltipKey, false, false, 1),
        },
        style: chartStyles.previousRequestData,
      },
      {
        childName: 'currentRequest',
        data: currentRequestData,
        legendItem: {
          name: getUsageRangeString(currentRequestData, requestKey, true, false),
          symbol: {
            fill: chartStyles.legendColorScale[3],
            type: 'dash',
          },
          tooltip: getUsageRangeString(currentRequestData, requestTooltipKey, false, false),
        },
        style: chartStyles.currentRequestData,
      },
    ];
    const cursorVoronoiContainer = this.getCursorVoronoiContainer();
    this.setState({ cursorVoronoiContainer, series });
  };

  private handleNavToggle = () => {
    setTimeout(this.handleResize, 500);
  };

  private handleResize = () => {
    if (this.containerRef.current) {
      this.setState({ width: this.containerRef.current.clientWidth });
    }
  };

  private getChart = (series: UsageChartSeries, index: number) => {
    const { hiddenSeries } = this.state;
    return (
      <ChartArea
        data={!hiddenSeries.has(index) ? series.data : [{ y: null }]}
        interpolation="monotoneX"
        key={series.childName}
        name={series.childName}
        style={series.style}
      />
    );
  };

  // Returns CursorVoronoiContainer component
  private getCursorVoronoiContainer = () => {
    // Note: Container order is important
    const CursorVoronoiContainer: any = createContainer('voronoi', 'cursor');

    return (
      <CursorVoronoiContainer
        cursorDimension="x"
        labels={this.getTooltipLabel}
        mouseFollowTooltips
        voronoiDimension="x"
        voronoiPadding={{
          bottom: 75,
          left: 8,
          right: 8,
          top: 8,
        }}
      />
    );
  };

  private getDomain() {
    const { series } = this.state;

    const domain: { x: DomainTuple; y?: DomainTuple } = { x: [1, 31] };
    let maxValue = 0;
    let minValue = 0;

    if (series) {
      series.forEach((s: any, index) => {
        if (!this.isSeriesHidden(index) && s.data && s.data.length !== 0) {
          const { max, min } = getMaxMinValues(s.data);
          maxValue = Math.max(maxValue, max);
          if (minValue === 0) {
            minValue = min;
          } else {
            minValue = Math.min(minValue, min);
          }
        }
      });
    }

    const threshold = maxValue * 0.1;
    const max = maxValue > 0 ? Math.ceil(maxValue + threshold) : 0;
    const _min = minValue > 0 ? Math.max(0, Math.floor(minValue - threshold)) : 0;
    const min = _min > 0 ? _min : 0;

    if (max > 0) {
      domain.y = [min, max];
    }
    return domain;
  }

  private getEndDate() {
    const { currentRequestData, currentUsageData, previousRequestData, previousUsageData } = this.props;
    const currentRequestDate = currentRequestData ? getDate(getDateRange(currentRequestData, true, true)[1]) : 0;
    const currentUsageDate = currentUsageData ? getDate(getDateRange(currentUsageData, true, true)[1]) : 0;
    const previousRequestDate = previousRequestData ? getDate(getDateRange(previousRequestData, true, true)[1]) : 0;
    const previousUsageDate = previousUsageData ? getDate(getDateRange(previousUsageData, true, true)[1]) : 0;

    return currentRequestDate > 0 || currentUsageDate > 0 || previousRequestDate > 0 || previousUsageDate > 0
      ? Math.max(currentRequestDate, currentUsageDate, previousRequestDate, previousUsageDate)
      : 31;
  }

  private getLegend = () => {
    const { legendItemsPerRow } = this.props;
    const { width } = this.state;

    // Todo: use PF legendAllowWrap feature
    const itemsPerRow = legendItemsPerRow ? legendItemsPerRow : width > 300 ? chartStyles.itemsPerRow : 1;

    return <ChartLegend data={this.getLegendData()} height={25} gutter={20} itemsPerRow={itemsPerRow} name="legend" />;
  };

  private getTooltipLabel = ({ datum }) => {
    const { formatDatumValue, formatDatumOptions } = this.props;
    const formatter = getTooltipContent(formatDatumValue);
    return datum.y !== undefined && datum.y !== null
      ? formatter(datum.y, datum.units, formatDatumOptions)
      : i18next.t('chart.no_data');
  };

  // Interactive legend

  // Hide each data series individually
  private handleLegendClick = props => {
    const hiddenSeries = new Set(this.state.hiddenSeries);
    if (!hiddenSeries.delete(props.index)) {
      hiddenSeries.add(props.index);
    }
    this.setState({ hiddenSeries });
  };

  // Returns true if at least one data series is available
  private isDataAvailable = () => {
    const { series } = this.state;
    const unavailable = []; // API data may not be available (e.g., on 1st of month)

    if (series) {
      series.forEach((s: any, index) => {
        if (this.isSeriesHidden(index) || (s.data && s.data.length === 0)) {
          unavailable.push(index);
        }
      });
    }
    return unavailable.length !== (series ? series.length : 0);
  };

  // Returns true if data series is hidden
  private isSeriesHidden = index => {
    const { hiddenSeries } = this.state; // Skip if already hidden
    return hiddenSeries.has(index);
  };

  private getAdjustedContainerHeight = () => {
    const { adjustContainerHeight, height, containerHeight = height } = this.props;
    const { width } = this.state;

    let adjustedContainerHeight = containerHeight;
    if (adjustContainerHeight) {
      if (width < 480) {
        adjustedContainerHeight += 20;
      }
    }
    return adjustedContainerHeight;
  };

  // Returns groups of chart names associated with each data series
  private getChartNames = () => {
    const { series } = this.state;
    const result = [];
    if (series) {
      series.map(serie => {
        // Each group of chart names are hidden / shown together
        result.push(serie.childName);
      });
    }
    return result as any;
  };

  // Returns onMouseOver, onMouseOut, and onClick events for the interactive legend
  private getEvents = () => {
    const result = getInteractiveLegendEvents({
      chartNames: this.getChartNames(),
      isHidden: this.isSeriesHidden,
      legendName: 'legend',
      onLegendClick: this.handleLegendClick,
    });
    return result;
  };

  // Returns legend data styled per hiddenSeries
  private getLegendData = (tooltip: boolean = false) => {
    const { hiddenSeries, series } = this.state;

    if (series) {
      const result = series.map((s, index) => {
        return {
          childName: s.childName,
          ...s.legendItem, // name property
          ...(tooltip && { name: s.legendItem.tooltip }), // Override name property for tooltip
          ...getInteractiveLegendItemStyles(hiddenSeries.has(index)), // hidden styles
        };
      });
      return result;
    }
  };

  public render() {
    const {
      height,
      padding = {
        bottom: 75,
        left: 8,
        right: 8,
        top: 8,
      },
      title,
    } = this.props;
    const { cursorVoronoiContainer, series, width } = this.state;

    const domain = this.getDomain();
    const endDate = this.getEndDate();
    const midDate = Math.floor(endDate / 2);

    // Clone original container. See https://issues.redhat.com/browse/COST-762
    const container = cursorVoronoiContainer
      ? React.cloneElement(cursorVoronoiContainer, {
          disable: !this.isDataAvailable(),
          labelComponent: (
            <ChartLegendTooltip
              legendData={this.getLegendData(true)}
              title={datum => i18next.t('chart.day_of_month_title', { day: datum.x })}
            />
          ),
        })
      : undefined;

    return (
      <>
        <Title headingLevel="h3" size="md">
          {title}
        </Title>
        <div className="chartOverride" ref={this.containerRef} style={{ height: this.getAdjustedContainerHeight() }}>
          <div style={{ height, width }}>
            <Chart
              containerComponent={container}
              domain={domain}
              events={this.getEvents()}
              height={height}
              legendComponent={this.getLegend()}
              legendData={this.getLegendData()}
              legendPosition="bottom-left"
              padding={padding}
              theme={ChartTheme}
              width={width}
            >
              {series &&
                series.map((s, index) => {
                  return this.getChart(s, index);
                })}
              <ChartAxis style={chartStyles.xAxis} tickValues={[1, midDate, endDate]} />
              <ChartAxis dependentAxis style={chartStyles.yAxis} />
            </Chart>
          </div>
        </div>
      </>
    );
  }
}

export { UsageChart, UsageChartProps };
