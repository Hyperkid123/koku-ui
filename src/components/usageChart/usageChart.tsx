import {
  ChartArea,
  ChartGroup,
  ChartLegend,
  ChartTheme,
  ChartVoronoiContainer,
} from '@patternfly/react-charts';
import { css } from '@patternfly/react-styles';
import {
  ChartDatum,
  getTooltipContent,
  getTooltipLabel,
} from 'components/commonChart/chartUtils';
import React from 'react';
import { FormatOptions, ValueFormatter } from 'utils/formatValue';
import { chartStyles, styles } from './usageChart.styles';

interface UsageChartProps {
  currentRequestData?: any;
  currentRequestLabel?: string;
  currentUsageData: any;
  currentUsageLabel?: string;
  formatDatumValue: ValueFormatter;
  formatDatumOptions?: FormatOptions;
  height: number;
  previousRequestData?: any;
  previousRequestLabel?: string;
  previousUsageData?: any;
  previousUsageLabel?: string;
  title?: string;
}

interface State {
  width: number;
}

class UsageChart extends React.Component<UsageChartProps, State> {
  private containerRef = React.createRef<HTMLDivElement>();
  public state: State = {
    width: 0,
  };

  public shouldComponentUpdate(nextProps: UsageChartProps) {
    if (
      !nextProps.currentUsageData ||
      !nextProps.previousUsageData ||
      !nextProps.currentRequestData ||
      !nextProps.previousRequestData
    ) {
      return false;
    }
    return true;
  }

  private getTooltipLabel = (datum: ChartDatum) => {
    const { formatDatumValue, formatDatumOptions } = this.props;
    return getTooltipLabel(
      datum,
      getTooltipContent(formatDatumValue),
      formatDatumOptions,
      'date'
    );
  };

  private handleResize = () => {
    this.setState({ width: this.containerRef.current.clientWidth });
  };

  public componentDidMount() {
    const node = this.containerRef.current;
    if (node) {
      this.setState({ width: node.clientWidth });
    }
    window.addEventListener('resize', this.handleResize);
  }

  public componentWillUnmount() {
    window.removeEventListener('resize', this.handleResize);
  }

  public render() {
    const {
      currentRequestData,
      currentRequestLabel,
      currentUsageData,
      currentUsageLabel,
      height,
      previousRequestData,
      previousRequestLabel,
      previousUsageData,
      previousUsageLabel,
      title,
    } = this.props;

    const firstRowLegendData = [];
    const secondRowLegendData = [];

    if (previousUsageData && previousUsageData.length) {
      firstRowLegendData.push({ name: previousUsageLabel });
    }
    if (previousRequestData && previousRequestData.length) {
      secondRowLegendData.push({ name: previousRequestLabel });
    }

    if (currentUsageData && currentUsageData.length) {
      firstRowLegendData.push({ name: currentUsageLabel });
    }
    if (currentRequestData && currentRequestData.length) {
      secondRowLegendData.push({ name: currentRequestLabel });
    }

    const container = <ChartVoronoiContainer labels={this.getTooltipLabel} />;

    return (
      <div className={css(styles.reportSummaryTrend)} ref={this.containerRef}>
        <div>
          <ChartGroup
            containerComponent={container}
            height={height}
            width={this.state.width}
          >
            {Boolean(currentUsageData && currentUsageData.length) && (
              <ChartArea
                style={chartStyles.currentUsageData}
                data={currentUsageData}
              />
            )}
            {Boolean(currentRequestData && currentRequestData.length) && (
              <ChartArea
                style={chartStyles.currentRequestData}
                data={currentRequestData}
              />
            )}
            {Boolean(previousUsageData && previousUsageData.length) && (
              <ChartArea
                style={chartStyles.previousUsageData}
                data={previousUsageData}
              />
            )}
            {Boolean(previousRequestData && previousRequestData.length) && (
              <ChartArea
                style={chartStyles.previousRequestData}
                data={previousRequestData}
              />
            )}
          </ChartGroup>
        </div>
        <ChartLegend
          title={title}
          theme={ChartTheme.dark.blue}
          colorScale={chartStyles.currentColorScale}
          data={firstRowLegendData}
          gutter={55}
          height={25}
          width={this.state.width}
        />
        <ChartLegend
          title={title}
          theme={ChartTheme.dark.blue}
          colorScale={chartStyles.previousColorScale}
          data={secondRowLegendData}
          height={25}
          width={this.state.width}
          style={{ data: { strokeDasharray: '5,5' } }}
        />
      </div>
    );
  }
}

export { UsageChart, UsageChartProps };
