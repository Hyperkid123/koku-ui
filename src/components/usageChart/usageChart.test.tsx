jest.mock('date-fns/format');

import { ChartArea, ChartBar, ChartGroup } from '@patternfly/react-charts';
import { OcpReport, OcpReportData } from 'api/ocpReports';
import * as utils from 'components/commonChart/chartUtils';
import formatDate from 'date-fns/format';
import { shallow } from 'enzyme';
import React from 'react';
import { UsageChart, UsageChartProps } from './usageChart';

const currentMonthReport: OcpReport = createReport('1-15-18');
const previousMonthReport: OcpReport = createReport('12-15-17');

const currentData = utils.transformOcpReport(
  currentMonthReport,
  utils.ChartType.daily,
  'usage'
);
const requestData = utils.transformOcpReport(
  previousMonthReport,
  utils.ChartType.daily,
  'request'
);

jest.spyOn(utils, 'getTooltipLabel');

const getTooltipLabel = utils.getTooltipLabel as jest.Mock;

const props: UsageChartProps = {
  title: 'Usage Title',
  height: 100,
  formatDatumValue: jest.fn(),
  currentData,
  formatDatumOptions: {},
  requestData,
};

test('reports are formatted to datums', () => {
  const view = shallow(<UsageChart {...props} />);
  const areaCharts = view.find(ChartArea);
  const barCharts = view.find(ChartBar);
  expect(areaCharts.length).toBe(1);
  expect(areaCharts.at(1).prop('data')).toMatchSnapshot('request data');
  expect(barCharts.length).toBe(1);
  expect(barCharts.at(1).prop('data')).toMatchSnapshot('current month data');
});

test('null previous and current reports are handled', () => {
  const view = shallow(
    <UsageChart {...props} currentData={null} requestData={null} />
  );
  const charts = view.find(ChartArea);
  expect(charts.length).toBe(0);
});

test('height from props is used', () => {
  const view = shallow(<UsageChart {...props} />);
  expect(view.find(ChartGroup).prop('height')).toBe(props.height);
});

test('labels formats with datum and value formatted from props', () => {
  const view = shallow(<UsageChart {...props} />);
  const datum: utils.ChartDatum = {
    x: 1,
    y: 1,
    key: '1-1-1',
    units: 'units',
  };
  const group = view.find(ChartGroup);
  group.props().containerComponent.props.labels(datum);
  expect(getTooltipLabel).toBeCalledWith(
    datum,
    props.formatDatumValue,
    props.formatDatumOptions,
    'date'
  );
  expect(props.formatDatumValue).toBeCalledWith(
    datum.y,
    datum.units,
    props.formatDatumOptions
  );
  expect(formatDate).toBeCalledWith(datum.key, expect.any(String));
  expect(view.find(ChartGroup).prop('height')).toBe(props.height);
});

test('labels ignores datums without a date', () => {
  const view = shallow(<UsageChart {...props} />);
  const datum: utils.ChartDatum = {
    x: 1,
    y: 1,
    key: '',
    units: 'units',
  };
  const group = view.find(ChartGroup);
  const value = group.props().containerComponent.props.labels(datum);
  expect(value).toBe('');
  expect(props.formatDatumValue).not.toBeCalled();
});

test('trend is a running total', () => {
  const multiDayReport: OcpReport = {
    data: [
      createReportDataPoint('1-15-18', 1),
      createReportDataPoint('1-16-18', 2),
    ],
  };
  const view = shallow(<UsageChart {...props} currentData={multiDayReport} />);
  const charts = view.find(ChartArea);
  expect(charts.at(1).prop('data')).toMatchSnapshot('current month data');
});

test('trend is a daily value', () => {
  const multiDayReport: OcpReport = {
    data: [
      createReportDataPoint('1-15-18', 1),
      createReportDataPoint('1-16-18', 2),
    ],
  };
  const view = shallow(<UsageChart {...props} currentData={multiDayReport} />);
  const charts = view.find(ChartArea);
  expect(charts.at(1).prop('data')).toMatchSnapshot('current month data');
});

function createReport(date: string): OcpReport {
  return {
    data: [createReportDataPoint(date)],
  };
}

function createReportDataPoint(date: string, usage = 1): OcpReportData {
  return {
    date,
    values: [{ date, usage, units: 'unit' }],
  };
}
