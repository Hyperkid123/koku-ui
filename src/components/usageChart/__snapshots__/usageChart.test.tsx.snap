// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reports are formatted to datums: current month data 1`] = `undefined`;

exports[`reports are formatted to datums: request data 1`] = `undefined`;

exports[`trend is a daily value: current month data 1`] = `undefined`;

exports[`trend is a running total: current month data 1`] = `undefined`;
